lib/prelude.dx

```diff
@@ -923,6 +873,218 @@ def isnan (x:Float) : Bool = not (x >= x && x <= x)
 -- Todo: use IEEE-754R 5.11: Floating Point Comparison Relation cmpUnordered.
 def either_is_nan (x:Float) (y:Float) : Bool = (isnan x) || (isnan y)
 
+'File system operations
```

Comment:
> It looks like a bigger change than it is because I had to manually re-toposort the prelude.

I wonder why manually reordering the prelude is necessary?

Reply:
We evaluate the file from top to bottom, just as if you'd typed it into the REPL. So you can't have function names appearing before they're defined. It's becoming enough of a pain that we should fix it.

Reply:
Ah, I see - declaration parsing is currently C-like. C and C++ have forward declarations (which are so hacky) to work around that problem.

Would a robust fix involve multi-phasic parsing? One phase to parse declaration types/prototypes, and a second phase to parse bodies - maybe more phases. What about declarations that don't have a declared type - type inference requires body parsing, so perhaps we could skip these declarations during phase one.

My experience with this comes from the [function prototype parsing phase from DLVM compiler's handwritten parser](https://github.com/dlvm-team/dlvm-core/blob/711067d3f2448cb7d104061de5db3e178de64ec5/Sources/DLParse/Parse.swift#L1594-L1622).